OpenRA.Mods.Common/Traits/Buildings/ProximityLinkHost.cs

```diff
@@ -0,0 +1,252 @@
+
+		[Desc("From how far away can clients be serviced?")]
+		public readonly WDist Range = WDist.FromCells(4);
+
+		[Desc("Does the client need to be preoccupied when linked?")]
+		public readonly bool OccupyClient = false;
+
+		public override object Create(ActorInitializer init) { return new ProximityLinkHost(init.Self, this); }
+	}
+
+	public class ProximityLinkHost : ConditionalTrait<ProximityLinkHostInfo>,
+		ILinkHost, ITick, INotifySold, INotifyOwnerChanged, ISync, INotifyKilled, INotifyActorDisposing
+	{
+		protected readonly Actor Self;
+
+		[Sync]
+		protected bool preventLink = false;
+
+		protected readonly List<LinkClientManager> ReservedLinkClients = new();
+		protected readonly List<(TraitPair<LinkClientManager> Client, long Time)> WaitingClients = new();
+		protected readonly List<TraitPair<LinkClientManager>> LinkedClients = new();
+		protected INotifyLinkHost[] notifyLinkHosts;
+
+		public ProximityLinkHost(Actor self, ProximityLinkHostInfo info)
+			: base(info)
+		{
+			Self = self;
+		}
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+			notifyLinkHosts = self.TraitsImplementing<INotifyLinkHost>().ToArray();
+		}
+
+		#region ILinkHost
+
+		public BitSet<LinkType> GetLinkType => Info.Type;
+
+		public bool IsEnabledAndInWorld => !preventLink && !IsTraitDisabled && !Self.IsDead && Self.IsInWorld;
+		public int ReservationCount => ReservedLinkClients.Count;
+		public bool CanBeReserved => Info.MaxQueueLength < 0 || ReservationCount < Info.MaxQueueLength;
+
+		public WPos LinkPosition => Self.CenterPosition + Info.LinkOffset;
+		public int LinkWait => Info.Wait;
+
+		public WAngle LinkFacing => WAngle.Zero;
+
+		public virtual bool IsLinkingPossible(Actor clientActor, ILinkClient client, bool ignoreReservations = false)
+		{
+			return !IsTraitDisabled && (ignoreReservations || CanBeReserved || ReservedLinkClients.Contains(client.LinkClientManager));
+		}
+
+		public virtual bool Reserve(Actor self, LinkClientManager client)
+		{
+			if (CanBeReserved && !ReservedLinkClients.Contains(client))
+			{
+				ReservedLinkClients.Add(client);
+				client.ReserveHost(self, this);
+				return true;
+			}
+
+			return false;
+		}
+
+		public virtual void UnreserveAll()
+		{
+			while (ReservedLinkClients.Count > 0)
+				Unreserve(ReservedLinkClients[0]);
+
+			WaitingClients.Clear();
+		}
+
+		public virtual void Unreserve(LinkClientManager client)
+		{
+			if (ReservedLinkClients.Remove(client))
+				client.UnreserveHost();
+		}
+
+		public virtual void OnLinkStarted(Actor self, Actor clientActor, LinkClientManager client)
+		{
+			if (Info.OccupyClient || LinkWait <= 0)
+				LinkStarted(self, new TraitPair<LinkClientManager>(clientActor, client));
+			else
+				WaitingClients.Add((new TraitPair<LinkClientManager>(clientActor, client), LinkWait + self.World.WorldTick));
+		}
+
+		public virtual void OnLinkCompleted(Actor self, Actor clientActor, LinkClientManager client)
```

Comment:
IMO the "started" and "completed" terms don't translate very well from Dock to Link (specifically, in this general case, "link completed" may imply that there should also be a "link failed"). Something like `OnLinkCreated` and `OnLinkRemoved` might be clearer?

Reply:
sure